search: Add support for case-sensitivity pattern items

dinocosta opened the pull request:
This Pull Request introduces support for pattern items in the buffer search. It does so by splitting the `query` methods into two new methods:

- `BufferSearchBar.raw_query` – returns the text from the search query editor
- `BufferSearchBar.query` - returns the search query with pattern items removed

Whenever the search query is updated, processing of the `EditorEvent::Edited` event ends up calling the `BufferSearchBar.apply_pattern_items` method, which parses the pattern items from the raw query, and updates the buffer search bar's search options accordingly. This `apply_pattern_items` function avoids updating the `BufferSearchBar.default_options` field in order to be able to reset the search options when a pattern items is removed. Lastly, new pattern items can easily be added by updating the `PATTERN_ITEMS` array.

### Screen Capture

https://github.com/user-attachments/assets/ebd83c38-e480-4c24-9b8c-6edde69cf392

---

Closes #32390

Release Notes:

- Added support for the `\c` and `\C` query pattern items to control case-sensitivity in buffer search

dinocosta committed:
feat(buffer_search): allow pattern items in search query

Update the way the search query in the buffer search bar is handled in
order to support pattern items. Right now only two pattern items are
supported:

- `\c` – When \c is provided in the search query, it disables the case
sensitivity search option
- `\C` – When \C is provided in the search query, it enables the case
sensitivity search option

This feature takes precedence over the `BufferSearchBar.search_options`,
but it ensures that, if any search option was previously
enabled/disabled and it was disabled/enabled by a pattern item, it'll
return to its initial state when the pattern item is deleted.

dinocosta committed:
refactor: use regex to remove pattern items from search query

This might be a bit of an overkill for the time being, but I expect that
when the array of pattern items grows, it might be faster to replace the
pattern items with a regex instead of iterating over the array.

dinocosta committed:
test(buffer_search): add test for pattern items

Add a very simple test that verifies the behaviour of
`apply_pattern_items` when the search query is updated. In practice it
would be best to be able to actually simulate the whole flow where
`EditorEvent::Edited` is triggered but I haven't managed to figure out
how to do that just yet.

dinocosta commented on crates/search/src/buffer_search.rs:
```diff
@@ -2931,6 +2994,105 @@ mod tests {
+            // Update the search query's text to `test\C` to check if the
+            // search option is correctly applied.
+            let query = "test\\C";
+            search_bar.query_editor.update(cx, |query_editor, cx| {
+                query_editor.buffer().update(cx, |query_buffer, cx| {
+                    let len = query_buffer.len(cx);
+                    query_buffer.edit([(0..len, query)], None, cx);
+                });
+            });
+
+            search_bar.apply_pattern_items(cx);
```
@ConradIrwin Unfortunately I wasn't able to get the test working with the suggested `cx.focus_view(&search_bar.query_editor)` and `cx.simulate_input("test\\C")`, it seems that might not actually trigger the editor event? 🤔 

Also tried both the `editor::Editor.handle_input` and `editor::Editor.set_text` methods, but both panic on `editor::Editor.text_layout_details`.

Let me know if there's any other approach you'd like me to try! And thank you for the help so far!

dinocosta committed:
fix: fix clippy issues

ConradIrwin committed:
Set up test properly

ConradIrwin commented:
It turns out that for that to work the view actually has to be rendered in the test window; fixed!

I'm not sure we need fancy_regex for this case (just regex will do), otherwise this seems reasonable to me.

dinocosta committed:
refactor: use regex over fancy-regex in pattern-items implementation

dinocosta commented:
> It turns out that for that to work the view actually has to be rendered in the test window; fixed!
> 
> I'm not sure we need fancy_regex for this case (just regex will do), otherwise this seems reasonable to me.

@ConradIrwin Thank you so much for taking a look and updating it; I would probably never have figured that one out 😅 

I've pushed a commit – https://github.com/zed-industries/zed/pull/34762/commits/83ca268b374dfd785cab9f80a39bd1e4fd99b510 – to replace the use of `fancy_regex` with `regex`. Do let me know if you think that's still a bit of overengineering, I'm also okay with simply updating `BufferSearchBar.query` to iterate over `PATTERN_ITEMS` and remove the pattern items from the string.

ConradIrwin commented:
🤷 Given that there are just two, maybe we can simplify all the code to be less general; but not too bothered by how it is currently.

One thing I missed in the first review: this can't be enabled for plain-text searches, we should only do it for regex searches (otherwise I can't search for \c).

dinocosta committed:
refactor: only apply pattern items if regex is enabled

dinocosta commented:
> 🤷 Given that there are just two, maybe we can simplify all the code to be less general; but not too bothered by how it is currently.
> 
> One thing I missed in the first review: this can't be enabled for plain-text searches, we should only do it for regex searches (otherwise I can't search for \c).

Thanks! I'll leave it as is for now 🙇 

Sorry for the late reply, was a bit busy these past days but I've just pushed [this commit](https://github.com/zed-industries/zed/pull/34762/commits/5c24007584370d37a0c5be20512b468b79a4aaba) to only apply the `\c` and `\C` pattern items if the `SearchOptions::REGEX` search option is enabled. The tests have also been updated to reflect this changes ✌️

dinocosta committed:
fix: fix issue with case-sensitivity turning off after first character

This commit fixes an issue with the pattern items implementation where,
in vim mode, deploying the buffer search through `/` and then typing any
character would automatically disable the regex setting. After some
research this was tracked down to the fact that the `default_options`
don't actually contain all of the settings that are enabled when the
buffer search bar is shown with `/`, as the vim layer is the one that
ends up calling the `set_search_options` method with the
`SearchOptions::REGEX` option enabled. Calling this method doesn't
actually update the `default_options`, and I believe this should
probably not be changed.

As such, in order to prevent this issue, this commit introduces a new
`pattern_item_options` vector to the `BufferSearchBar`, which simply
keeps track of the search options that were either enabled/disabled by
pattern items in the search query, that actually had impact on the
search options. By keeping track of this, one can easily just revert the
options, by calling `toggle_search_option`, seeing as we only keep track
of the ones that actually had any effect on the options, and revert to
all of the options that were enabled when the buffer search bar was
deployed, be it on the `default_options` or not.

I believe the current implementation can probably be improved with a
simple `SearchOptions` and then using the `exclusion` or `intersection`
methods but have yet to dedicate more time to this. On the other hand,
there's yet another issue that surfaced while working on this:

1. Type `Main\c\C\c` in the search query 2. Confirm that
case-sensitivity is enabled

This happens because we're not actually keeping track of the order the
pattern items are applied in the search query, so the order they're
defined in `PATTERN_ITEMS` takes precendence, we'll likely need to
capture the matches so we can then actually leverage the order to
determine what the final option should be.

dinocosta committed:
fix: fix issue with conflicting pattern items

Fix issue where having conflicting pattern items in the search query
could lead to erroneous results. For example, if the search query was
`test\c\C\c`, one would expect for the case sensitivity option to be
disabled, seeing as `\c` is the last pattern item. However, since the
code was simply iterating over `PATTERN_ITEMS`, it did not take into
consideration that:

1. The same pattern item could appear more than once in the query 2. The
order of the pattern item in the query should take precedence over the
order of the pattern item in `PATTERN_ITEMS`

As such, this commit fixes that so that the implementation actually
iterates over the pattern items capture in the query, ensuring that, if
`\c\C\c` is found, the last pattern item is the one that will be
applied.

dinocosta committed:
fix: fix issue with \ preceding pattern items

With the current implementation of pattern items, it's impossible to
search for strings like `\c` in buffers as, even when using `\\c`, the
`\c` suffix will be interpreted as a pattern item. This commit updates
the regex used to find the pattern items so as to ensure that the
preceding character is never a `\` and, if it is, it's considered that
the user is searching for a slash instead of trying to use a pattern
item.

dinocosta committed:
refactor: reimplement pattern items as an enum

Update the way pattern items are implemented so that, instead of saving
these in an array with the character representation, the search option
and its value, it now is a `PatternItem` enum, and the enum implements
several methods:

- `character` - Returns the character representation for the pattern
item item, without the backslash.
- `search_option` - Returns the search option and its desired value.

The `TryFrom<&str>` trait is also implemented for `PatternItem` so that
we can simply build a `PatternItem` from a string like `"\\c"`.

dinocosta committed:
refactor: introduce pattern_items module

dinocosta committed:
feat: add pattern items to project search

Update the `search::project_search::ProjectSearchView` implementation in
order to allow users to leverage pattern items in the search query.

Add a very basic `test_pattern_items` test to `search::project_search`
in order to test how the `apply_pattern_items` method affects the
`ProjectSearchView.search_options` value depending on the query editor's
text. Unfortunately I wasn't having much luck adding a text similar to
the one for `BufferSearchBar` where we actually simulate the user's
input and keystrokes, so definitely something that can be improved upon.

dinocosta committed:
refactor: implement case sensitive pattern items in search query

In order to simplify the implementation of pattern items in search
queries, this commit updates the `project::search::SearchQuery::regex`
function so as to support both `\\c` and `\\C` in the provided query.

This means that we no longer need to have both `BufferSearchBar` and
`ProjectSearchView` handling pattern items, so the
`search::pattern_items` module can now safely be removed.

It's worth noting that, since these are now handled at the `SearchQuery`
level, this removes the updates to the UI regarding search options, that
were being triggered by the pattern items being processed and applied to
the search options.

Co-authored-by: Conrad Irwin <conrad.irwin@gmail.com>

dinocosta committed:
chore(diagnostics): remove pattern_items module

With the previous commit, both `BufferSearchBar` and `ProjectSearchView`
have been updated to no longer rely on `PatternItems`, seeing as
`SearchQuery::regex` is now responsible for handling pattern items.

This commit removes the `search::pattern_items` module, since there's no
longer other modules needing its functionality.

ConradIrwin commented:
Thanks!

Are the remaining changes in the views needed, or are they just hold-overs?

We should probably write a search page... No need to do that before we merge this, but it would be good!

dinocosta committed:
chore: cleanup unnecessary comments and unused method

- Remove comments from both `BufferSearchBar.query` and
`ProjectSearchView.query` mentioning pattern items
- Remove unused `BufferSearchBar.disable_search_option` method
